src/common/AdaptiveStream.cpp

```diff
@@ -840,37 +841,23 @@ bool AdaptiveStream::ensureSegment()
 
       size_t nextsegmentPosold = current_rep_->get_segment_pos(nextSegment);
       uint64_t nextsegno = current_rep_->getSegmentNumber(nextSegment);
-      CRepresentation* newRep{nullptr};
 
-      bool lastSeg{false};
-      if (current_period_ != tree_.m_periods.back().get())
-      {
-        if (nextsegmentPosold + available_segment_buffers_ ==
-            current_rep_->SegmentTimeline().GetSize() - 1)
-        {
-          lastSeg = true;
-        }
-      }
-      
-      if (segment_buffers_[0]->segment.IsInitialization() ||
-          valid_segment_buffers_ == 0 ||
-          current_adp_->GetStreamType() != StreamType::VIDEO)
-      {
-        newRep = current_rep_;
-      }
-      else if (lastSeg) // Don't change reps on last segment of period, use the rep of preceeding seg
-      {
-        newRep = segment_buffers_[valid_segment_buffers_ - 1]->rep;
-      }
-      else
+      CRepresentation* newRep = current_rep_;
+      bool isBufferFull = valid_segment_buffers_ >= max_buffer_length_;
+
+      if (!segment_buffers_[0]->segment.IsInitialization() && available_segment_buffers_ > 0 &&
+          !isBufferFull) // Defer until we have some free buffer
       {
-        // Defer until we have some free buffer
-        if (available_segment_buffers_ < max_buffer_length_) {
-          newRep = tree_.GetRepChooser()->GetNextRepresentation(
-              current_adp_, segment_buffers_[available_segment_buffers_ - 1]->rep);
-        }
+        // The representation from the last added segment buffer
+        CRepresentation* prevRep = segment_buffers_[available_segment_buffers_ - 1]->rep;
+
+        bool isLastSegment = nextsegmentPosold + available_segment_buffers_ ==
+                             current_rep_->SegmentTimeline().GetSize() - 1;
+        // Dont change representation if it is the last segment of a period
+        if (isLastSegment)
+          newRep = prevRep;
```

Comment:
I forgot about this.

I don't think my existing code was correct because I noticed a crash on this once. It was a long time ago so I can't remember the exact details - hoping that your change from using valid_segment_buffers_ to available_segment_buffers_ is where my error was.

FYI the reason behind adding this condition was that when it comes time to play the last segment in a period, AdaptiveStream wasn't able to insert the initialization segment (in the case of fMP4) and you would get corrupted or blank video for the last segment.

Reply:
hmm better if i add this as comment in the code